### PR TITLE
[pkg/quantile] Add function to translate DDSketches into Sketches

### PR DESCRIPTION
**Pull request by KSerrania**
### What does this PR do?

Adds a way to convert an input DDSketch into a Sketch that can be sent to the backend.

### Motivation

Needed to support ExponentialHistograms in the OpenTelemetry collector's Datadog exporter.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/quantile/ddsketch.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	negativeStore := store.NewDenseStore()
+
+	gamma := c.gamma.v
+	// TODO: using 1338 instead of 1338.5 gave better results for DDSketch -> Sketch conversion
+	//       Why? May have to do with the way we convert float counts to int counts in
+	//       keyCountsFromFloatKeyCounts.
+	offset := float64(c.norm.bias)
+	newMapping, err := mapping.NewLogarithmicMappingWithGamma(gamma, offset)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create LogarithmicMapping for DDSketch: %v", err)
```
Comment (mx-psi) [inline pkg/quantile/ddsketch.go:39]: Use `%w` for a [wrapped error](https://pkg.go.dev/errors#pkg-overview)
```suggestion
		return nil, fmt.Errorf("couldn't create LogarithmicMapping for DDSketch: %w", err)
```

File: pkg/quantile/ddsketch.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	// Calculate the total count that was inserted in the Sketch
+	var cnt uint
+	for _, v := range keyCounts {
+		cnt += v.n
+	}
+	sum := inputSketch.GetSum()
+	avg := sum / float64(cnt)
+	max, err := inputSketch.GetValueAtQuantile(1.0)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't compute maximum of ddsketch: %v", err)
```
Comment (mx-psi) [inline pkg/quantile/ddsketch.go:166]:
```suggestion
		return nil, fmt.Errorf("couldn't compute maximum of ddsketch: %w", err)
```

File: pkg/quantile/ddsketch.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	min, err := inputSketch.GetValueAtQuantile(0.0)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't compute minimum of ddsketch: %v", err)
```
Comment (mx-psi) [inline pkg/quantile/ddsketch.go:171]:
```suggestion
		return nil, fmt.Errorf("couldn't compute minimum of ddsketch: %w", err)
```

File: pkg/quantile/ddsketch.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	return outputSketch, nil
+}
+
+// FromDDSketch converts a DDSketch into a Sketch
+func FromDDSketch(inputSketch *ddsketch.DDSketch) (*Sketch, error) {
+	sketchConfig := Default()
+
+	compatibleDDSketch, err := convertToCompatibleDDSketch(sketchConfig, inputSketch)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't convert input ddsketch into ddsketch with compatible parameters: %v", err)
```
Comment (mx-psi) [inline pkg/quantile/ddsketch.go:197]:
```suggestion
		return nil, fmt.Errorf("couldn't convert input ddsketch into ddsketch with compatible parameters: %w", err)
```

File: pkg/quantile/ddsketch.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	outputSketch, err := fromCompatibleDDSketch(sketchConfig, compatibleDDSketch)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't convert ddsketch into Sketch: %v", err)
```
Comment (mx-psi) [inline pkg/quantile/ddsketch.go:202]:
```suggestion
		return nil, fmt.Errorf("couldn't convert ddsketch into Sketch: %w", err)
```

File: pkg/quantile/ddsketch_test.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	// Support of the distribution: [0,N]
+	N := 1_000
+	// Number of points per quantile
+	M := 50
+
+	tests := []struct {
+		// distribution name
+		name string
+		// the cumulative distribution function (within [0,N])
+		cdf func(x float64) float64
```
Comment (mx-psi) [inline pkg/quantile/ddsketch_test.go:42]: we don't really use the `cdf` function here, it can be removed

File: pkg/quantile/ddsketch_test.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
```
Comment (mx-psi) [inline pkg/quantile/ddsketch_test.go:143]: ditto

File: pkg/quantile/ddsketch.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+func createDDSketchWithSketchMapping(c *Config, inputSketch *ddsketch.DDSketch) (*ddsketch.DDSketch, error) {
+	// Create positive store for the new DDSketch
+	positiveStore := store.NewDenseStore()
+
+	// Create negative store for the new DDSketch
+	negativeStore := store.NewDenseStore()
+
+	// Take parameters that match the Sketch mapping, and create a LogarithmicMapping out of them
+	gamma := c.gamma.v
+	offset := float64(c.norm.bias) + 0.5
```
Comment (CharlesMasson) [inline pkg/quantile/ddsketch.go:34]: Nit.: for future reference, maybe add a comment to explain the 0.5 shift (taking floor value in DDSketch vs. rounding to nearest integer in the agent sketch).

**Review by mx-psi — APPROVED** [review]
LGTM (once sketches-go version is updated)

**Review by CharlesMasson — APPROVED** [review]
LGTM.

**Pull request merged**